### ISSUE TITLE
arrow dataset: how to use date.year and date.month as partitioning

### ISSUE THREAD

**Issue by code1704** [apache/arrow#14619]
The data columns:
datetime, value

I want to build a dataset with partitioning like ("date.year", "date.month"). How to do that?
And when do filtering like date > "2021-09-02" and date < "2022-04-06", will it read the partition files between 2021-09 to 2022-04 only?

**Comment by djouallah** [comment]
yes, we do have it in BigQuery and it is freaking amazing !!

**Comment by code1704** [comment]
Thanks @djouallah . What is BigQuery? Is it a feature of duckdb?

**Comment by westonpace** [comment]
Are you using pyarrow?  Have you looked at https://arrow.apache.org/docs/python/dataset.html#writing-partitioned-data ?

> And when do filtering like date > "2021-09-02" and date < "2022-04-06", will it read the partition files between 2021-09 to 2022-04 only?

Yes, that should work.

**Comment by djouallah** [comment]
@westonpace I think we are talking about two different thing, ideally arrow should be able to partition by a field date, but instead of generating a file by day, it will generate a file by year, I don't think it is supported yet, a workaround is to create a new field year and use it as a partition column, the problem is you have to add it as a filter in the where clause.

BigQuery and I think iceberg support that functionality partition by Year (field)

**Comment by westonpace** [comment]
Thank you for the clarification.  You are correct that this is not supported.  Just to clarify.  The directories are structured something like:

```
/year=2020/month=01/day=01/chunk-0.parquet
...
/year=2021/month=03/day=12/chunk-0.parquet
```

Also, it sounds like there will be cases where you do not have the day (or maybe that is the typical case), correct?

**Comment by djouallah** [comment]
@westonpace I don't know how it is implemented, BigQuery don't expose its file format.

**Comment by code1704** [comment]
> Thank you for the clarification. You are correct that this is not supported. Just to clarify. The directories are structured something like:
> 
> ```
> /year=2020/month=01/day=01/chunk-0.parquet
> ...
> /year=2021/month=03/day=12/chunk-0.parquet
> ```
> 
> Also, it sounds like there will be cases where you do not have the day (or maybe that is the typical case), correct?

Yes. And we use duckdb as the query engine: duckdb + arrow dataset. For query like `SELECT * FROM some_dataset WHERE date BETWEEN "2015-03-05" AND "2018-04012"`, partition pruning is expected to work.

**Comment by jmakov** [comment]
A bit surprised how this isn't a bigger issue. Doesn't anybody use arrow partitions or do they just use other solutions?

**Comment by steinwaywhw** [comment]
Running into the same situations. 

The issue is that year, month, day, when used as partition columns, can't be compared independently of each other when writing filters for push down. They should be combined. 

Does Arrow allow things like "make_a_date(year, month, day) > a_date" to be pushed down?

**Comment by AlexisBRENON** [comment]
I endend up with such kind of solutions:

```python
import pyarrow as pa
from pyarrow import compute as pc, dataset as ds, fs, parquet as pq

partition_timestamp = pc.strptime(  # type: ignore
    pc.binary_join_element_wise(  # type: ignore
        pc.field("year"),
        pc.field("month"),
        pc.field("day"),
        pc.field("hour"),
        pc.scalar("+0000"),
        "-",
    ),
    format="%Y-%m-%d-%H-%z",
    unit="us",
)

# Load any partiton overlapping the requested interval. All intervals are open-ended ([start, end[).
# A partition interval is [P, P+1H[. It overlap the requested interval [dis, die[ if:
#   cond1: it is not before -> ! (P+1H <= dis)
#   cond2: it is not after -> ! (die <= P)
# So: overlap = cond1 && cond2
#             = ! (P+1H <= dis) && ! (die <= P)
#             = ! (P+1H <= dis || die <= P)  (De Morgan's law)
#             = ! (P <= dis - 1H  || die <= P)
# https://stackoverflow.com/a/325964
partition_filters = ~(
    (partition_timestamp <= data_interval_start - datetime.timedelta(hours=1))
    | (data_interval_end <= partition_timestamp)
)

return pd.DataFrame(
    pq.ParquetDataset(
        root_folder,
        filters=partition_filters,
    )
    .read_pandas()
    .to_pandas()
)
```

However, I remember of a performance issue on cloud blob store (too many listing or something like this) and finally rolled back to a path generation routine:

```python
for ts in pd.date_range(
    data_interval_start,
    data_interval_end,
    freq="1H",
    inclusive="left",
).tolist():
  partition = f"year={ts.year}/month={ts.month:02d}/day={ts.day:02d}/hour={ts.hour:02d}/"
  yield pd.read_parquet(
    f"{root_folder}/{partition}",
  )
```

But this seems less generic and fails to handle other partition-columns...
